File: benchmarks/pandas/bench_dataframe_head_tail.py
"""Benchmark: DataFrame.head() and .tail() — slice first/last N rows."""
import json, time
import pandas as pd

SIZE = 100_000
WARMUP = 5
ITERATIONS = 50

df = pd.DataFrame({"a":[float(i) for i in range(SIZE)],"b":[i*2 for i in range(SIZE)],"c":[str(i) for i in range(SIZE)]})

for _ in range(WARMUP):
    df.head(100)
    df.tail(100)

times = []
for _ in range(ITERATIONS):
    t0 = time.perf_counter()
    df.head(100)
    df.tail(100)
    times.append((time.perf_counter() - t0) * 1000)

total_ms = sum(times)
print(json.dumps({"function":"dataframe_head_tail","mean_ms":round(total_ms/ITERATIONS,3),"iterations":ITERATIONS,"total_ms":round(total_ms,3)}))

File: benchmarks/tsb/bench_dataframe_head_tail.ts
/**
 * Benchmark: DataFrame.head() and .tail() — slice first/last N rows.
 * Outputs JSON: {"function": "dataframe_head_tail", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import { DataFrame } from "../../src/index.ts";

const SIZE = 100_000;
const WARMUP = 5;
const ITERATIONS = 50;

const df = new DataFrame({
  a: Array.from({ length: SIZE }, (_, i) => i * 1.0),
  b: Array.from({ length: SIZE }, (_, i) => i * 2),
  c: Array.from({ length: SIZE }, (_, i) => String(i)),
});

for (let i = 0; i < WARMUP; i++) {
  df.head(100);
  df.tail(100);
}

const times: number[] = [];
for (let i = 0; i < ITERATIONS; i++) {
  const start = performance.now();
  df.head(100);
  df.tail(100);
  times.push(performance.now() - start);
}

const totalMs = times.reduce((a, b) => a + b, 0);
const meanMs = totalMs / ITERATIONS;
console.log(
  JSON.stringify({
    function: "dataframe_head_tail",
    mean_ms: Math.round(meanMs * 1000) / 1000,
    iterations: ITERATIONS,
    total_ms: Math.round(totalMs * 1000) / 1000,
  }),
);
